atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/ShardProgress.java

```diff
@@ -126,29 +126,49 @@ private long increaseValueFromToAtLeast(ShardAndStrategy shardAndStrategy, long
                 return newVal;
             } catch (CheckAndSetException e) {
                 log.info("Failed to check and set from expected old value {} to new value {}. Retrying if the old "
-                        + "value changed under us.",
-                        SafeArg.of("old value", oldVal),
+                                + "value changed under us.",
+                        SafeArg.of("old value", currentValue),
                         SafeArg.of("new value", newVal));
-                currentValue = updateOrRethrowIfNoChange(shardAndStrategy, currentValue, e);
+                currentValue = rethrowIfUnchanged(shardAndStrategy, currentValue, e);
             }
         }
         return currentValue;
     }
 
-    private long updateOrRethrowIfNoChange(ShardAndStrategy shardAndStrategy, long oldVal, CheckAndSetException ex) {
-        long updatedOldVal = getValue(getEntry(shardAndStrategy));
+    static byte[] createColumnValue(long newVal) {
+        return SweepShardProgressTable.Value.of(newVal).persistValue();
+    }
+
+    private long rethrowIfUnchanged(ShardAndStrategy shardStrategy, long oldVal, CheckAndSetException ex) {
+        long updatedOldVal = getValue(getEntry(shardStrategy));
         if (updatedOldVal == oldVal) {
             throw ex;
         }
         return updatedOldVal;
     }
 
     private CheckAndSetRequest createRequest(ShardAndStrategy shardAndStrategy, long oldVal, byte[] colValNew) {
-        if (oldVal == SweepQueueUtils.INITIAL_TIMESTAMP
-                || (shardAndStrategy == SHARD_COUNT_SAS && oldVal == AtlasDbConstants.DEFAULT_SWEEP_QUEUE_SHARDS)) {
-            return CheckAndSetRequest.newCell(TABLE_REF, cellForShard(shardAndStrategy), colValNew);
+        if (isDefaultValue(shardAndStrategy, oldVal)) {
+            return maybeGet(shardAndStrategy)
+                    .map(persistedValue -> createSingleCellRequest(shardAndStrategy, persistedValue, colValNew))
+                    .orElse(createNewCellRequest(shardAndStrategy, colValNew));
+        } else {
+            return createSingleCellRequest(shardAndStrategy, oldVal, colValNew);
         }
-        byte[] colValOld = SweepShardProgressTable.Value.of(oldVal).persistValue();
+    }
+
+    private boolean isDefaultValue(ShardAndStrategy shardAndStrategy, long oldVal) {
+        return oldVal == SweepQueueUtils.INITIAL_TIMESTAMP
```

Comment:
Comment here about the bug? Comparing the number of shards to a timestamp reads strangely